### ISSUE TITLE
Different results from xgboost when using daal4py converted model

### ISSUE THREAD

**Issue by davechallis** [intel/scikit-learn-intelex#1819]
When I convert an XGBoost model into a daal4py model and run inference using it, I get different results for the same input data.

My code is for a binary classifier, roughly something like:

```python
import pickle
import xgboost as xgb
import daal4py as d4p

model_xgb = xgb.Booster().load_model("model.bin")

model_d4p = d4p.convert_model(xgb.Booster().load_model("model.bin"))

with open("data.pkl", "rb") as fh:
    X = pickle.load(fh)

print(f"X type={type(X)}, datatype={X.dtype}, shape={X.shape}")

y_xgb = model_xgb.predict(xgb.DMatrix(X))

y_d4p = model_d4p.predict_proba(X)[:, 0]

print("xgb positives:", (y_xgb >= 0.5).sum()
print("d4p positives:", (y_d4p >= 0.5).sum()
```
This outputs:
```
X type=<class 'scipy.sparse._csr.csr_matrix'>, datatype=float32, shape=(80000, 5300)
xgb positives: 62
d4p positives: 1308
```
Whereas I would have expected both predictions to return identical results. Is there anything I've done that's obviously wrong here, or something you'd suggest to investigate further?

I can see if I can generate a smaller sample of test data if that's useful.

In case it helps, these are my model's parameters, extracted using `d4p.get_xgboost_params(bst)`:
```python
{'learner': {'generic_param': {'device': 'cpu',
   'fail_on_invalid_gpu_id': '0',
   'n_jobs': '1',
   'nthread': '1',
   'random_state': '0',
   'seed': '0',
   'seed_per_iteration': '0',
   'validate_parameters': '1'},
  'gradient_booster': {'gbtree_model_param': {'num_parallel_tree': '1',
    'num_trees': '100'},
   'gbtree_train_param': {'process_type': 'default',
    'tree_method': 'auto',
    'updater': 'grow_quantile_histmaker',
    'updater_seq': 'grow_quantile_histmaker'},
   'name': 'gbtree',
   'specified_updater': False,
   'tree_train_param': {'alpha': '0',
    'cache_opt': '1',
    'colsample_bylevel': '1',
    'colsample_bynode': '1',
    'colsample_bytree': '1',
    'eta': '0.300000012',
    'gamma': '0',
    'grow_policy': 'depthwise',
    'interaction_constraints': '',
    'lambda': '1',
    'learning_rate': '0.300000012',
    'max_bin': '256',
    'max_cat_threshold': '64',
    'max_cat_to_onehot': '4',
    'max_delta_step': '0',
    'max_depth': '6',
    'max_leaves': '0',
    'min_child_weight': '1',
    'min_split_loss': '0',
    'monotone_constraints': '()',
    'refresh_leaf': '1',
    'reg_alpha': '0',
    'reg_lambda': '1',
    'sampling_method': 'uniform',
    'sketch_ratio': '2',
    'sparse_threshold': '0.20000000000000001',
    'subsample': '1'},
   'updater': [{'hist_train_param': {'debug_synchronize': '0',
      'max_cached_hist_node': '65536'},
     'name': 'grow_quantile_histmaker'}]},
  'learner_model_param': {'base_score': '5E-1',
   'boost_from_average': '1',
   'num_class': '0',
   'num_feature': '5300',
   'num_target': '1'},
  'learner_train_param': {'booster': 'gbtree',
   'disable_default_eval_metric': '0',
   'multi_strategy': 'one_output_per_tree',
   'objective': 'binary:logistic'},
  'metrics': [],
  'objective': {'name': 'binary:logistic',
   'reg_loss_param': {'scale_pos_weight': '1'}}},
 'version': [2, 0, 3]}
```

I've tested on:
OS: Mac OS X, Ubuntu 22.04
Python: 3.11, 3.12
dall4py: 2024.3.0 on linux, daal4py==2023.2.1 on mac os x
xgboost: 2.0.3

Happy to provide more info, and follow up on any suggestions.

**Comment by razdoburdin** [comment]
Hi, 
thanks for using daal4py!

I didn't catch from your example, why did you use `predict` for xgboost and `predict_proba` for d4p? These two methods are not equivalent to each over.

**Comment by davechallis** [comment]
@razdoburdin Hi, thanks for taking a look! I'm using the xgboost [python API predict](https://xgboost.readthedocs.io/en/stable/python/python_api.html#xgboost.Booster.predict) and not the scikit-learn wrapper, so the xgboost `predict` function returns probabilities rather than labels (unless I'm misunderstanding the API).

E.g. in the example above, if I print out the first 5 results in each `y`:
```python
y_xgb = model_xgb.predict(xgb.DMatrix(X))
y_d4p = model_d4p.predict_proba(X)[:, 0]

print("xgb", y_xgb[:5])
print("d4p", y_d4p[:5])
```
this outputs:
```
xgb [0.00546392 0.00123668 0.00123668 0.00136322 0.00360578]
d4p [0.3883016  0.38513577 0.38620013 0.38620013 0.32219833]
```

**Comment by davechallis** [comment]
@razdoburdin Just to make it a clearer comparison, I also did a quick check with using the sklearn API and outputting the full results of `predict_proba` in both cases:
```python
import pickle
import xgboost as xgb
import daal4py as d4p

model_xgb_sklearn = xgb.XGBClassifier()
model_xgb_sklearn.load_model("model.bin")

model_d4p = d4p.convert_model(xgb.Booster().load_model("model.bin"))

with open("data.pkl", "rb") as fh:
    X = pickle.load(fh)

y_xgb = model_xgb_sklearn.predict_proba(X)
y_d4p = model_d4p.predict_proba(X)

print("xgb")
print(y_xgb[:5])
print()
print("d4p")
print(y_d4p[:5])
```
which outputs:
```
xgb
[[0.9945361  0.00546392]
 [0.9987633  0.00123668]
 [0.9987633  0.00123668]
 [0.9986368  0.00136322]
 [0.9963942  0.00360578]]

d4p
[[0.3883016  0.6116984 ]
 [0.38513577 0.61486423]
 [0.38620013 0.61379987]
 [0.38620013 0.61379987]
 [0.32219833 0.67780167]]
```

**Comment by razdoburdin** [comment]
Hi @davechallis ,
I wasn't able to reproduce the problem.
Could you please provide some launchable reproducer?

**Comment by davechallis** [comment]
@razdoburdin no problem, give me a while to sample some data, and I'll try and get someone that works end to end uploaded here.

**Comment by razdoburdin** [comment]
Hi @davechallis ,
thanks for the reproducer. The problem you found is caused by incorrect working with missing values for scipy csr matrices.
As a quick fix you can add to your code the following:
```
X_np = np.empty(X.shape)
X_np.fill(np.nan)
X_np[X.nonzero()] = X[X.nonzero()]
y_d4p = model_d4p.predict_proba(X_np)
```
![image](https://github.com/intel/scikit-learn-intelex/assets/23584552/6aba7442-a887-4c6d-bbac-0a12cee4de60)

Let me know if found any other problems!

**Comment by davechallis** [comment]
@razdoburdin Fantastic, that works perfectly! Thanks for your help, I've tested with my full example, and all works great.

**Comment by davechallis** [comment]
@razdoburdin One last thing I thought I'd ask/check - is there any other approach for this that doesn't involve converting to a dense matrix first? I've tested on a few hundred classifiers I've got, and some of them are extremely sparse, so I end up hitting memory issues when converting to a dense matrix.

Or if there's any documentation/source I can read to find out more, that'd also be great.

If not, then no problem, I can maybe check matrix density then swap between daal4py and native xgboost models depending on that.
